dbfit-java/hsqldb/src/main/java/dbfit/environment/HSQLDBEnvironment.java

```diff
@@ -65,47 +68,99 @@ protected Pattern getParameterPattern() {
+            dc.setString(2, schemaName);
             ResultSet rs = dc.executeQuery();
             Map<String, DbParameterAccessor> allParams = new HashMap<String, DbParameterAccessor>();
             int position = 0;
             while (rs.next()) {
-                String columnName = rs.getString(1);
-                String dataType = rs.getString(2);
+                String columnName = rs.getString("PARAMETER_NAME");
+                String dataType = rs.getString("TYPE_NAME");
+                Direction direction = null;
+                switch (rs.getString("PARAMETER_MODE")) {
+                    case "IN":
+                        direction = Direction.INPUT;
+                        break;
+                    case "INOUT":
+                        direction = Direction.INPUT_OUTPUT;
+                        break;
+                    case "OUT":
+                        direction = Direction.OUTPUT;
+                        break;
+                    case "RETURN_VALUE":
+                        direction = Direction.RETURN_VALUE;
+                        break;
+                }
                 DbParameterAccessor dbp = createDbParameterAccessor(
                         columnName,
-                        INPUT,
+                        direction,
                         typeMapper.getJDBCSQLTypeForDBType(dataType),
-                        getJavaClass(dataType), position++);
+                        getJavaClass(dataType), position++ - 1);
                 allParams.put(NameNormaliser.normaliseName(columnName), dbp);
             }
             rs.close();
             return allParams;
         }
     }
 
-    public Map<String, DbParameterAccessor> getAllProcedureParameters(String s)
+    public Map<String, DbParameterAccessor> getAllProcedureParameters(String routineName)
             throws SQLException {
-        throw new UnsupportedOperationException();
+        String qry = "SELECT parameter_name"
+                   + "           AS column_name"
+                   + "     , data_type"
+                   + "           AS type_name"
+                   + "     , parameter_mode"
+                   + "  FROM ("
+                   + "       SELECT rout.routine_schema"
+                   + "            , rout.routine_name"
+                   + "            , para.parameter_name"
+                   + "            , para.data_type"
+                   + "            , para.parameter_mode"
+                   + "            , para.ordinal_position"
+                   + "         FROM information_schema.routines rout"
+                   + "        INNER"
+                   + "         JOIN information_schema.parameters para"
+                   + "           ON rout.specific_catalog = para.specific_catalog"
+                   + "          AND rout.specific_schema = para.specific_schema"
+                   + "          AND rout.specific_name = para.specific_name"
+                   + "        UNION"
```

Comment:
To be it's not a clear cut case which is more natural. If `UNION ALL` were defined as 'do not remove duplicates' then it might suggest that we have them and we want to allow them.

If we were to use `UNION` then it might suggest that we have, or could have, duplicates and that we want them removed.

What's your view?